sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+
+    private Read spec;
+
+    public UnboundedMqttSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public UnboundedReader createReader(PipelineOptions options,
+                                        CheckpointMark checkpointMark) {
+      return new UnboundedMqttReader(this);
+    }
+
+    @Override
+    public List<UnboundedMqttSource> generateInitialSplits(int desiredNumSplits,
+                                                           PipelineOptions options) {
+      List<UnboundedMqttSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        // NB: it's important that user understand the impact of MQTT message QoS
+        sources.add(new UnboundedMqttSource(spec));
+      }
+      return sources;
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return VoidCoder.of();
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(byte[].class);
+    }
+
+  }
+
+  private static class UnboundedMqttReader extends UnboundedSource.UnboundedReader<byte[]> {
+
+    private final UnboundedMqttSource source;
+
+    private transient MqttClient client;
+    private transient byte[] current;
+    private transient Instant currentTimestamp;
+    private transient Instant watermark;
+    private transient BlockingQueue<KV<MqttMessage, Instant>> queue;
+
+    private UnboundedMqttReader(UnboundedMqttSource source) {
+      this.source = source;
+      this.current = null;
+      this.queue = new LinkedBlockingQueue<>();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOGGER.info("Starting MQTT reader");
+      Read spec = source.spec;
+      try {
+        client = spec.mqttConnectionConfiguration().getClient();
+        client.setCallback(new MqttCallback() {
+          @Override
+          public void connectionLost(Throwable cause) {
+            LOGGER.warn("MQTT connection lost", cause);
+          }
+
+          @Override
+          public void messageArrived(String topic, MqttMessage message) throws Exception {
+            LOGGER.info("Message arrived");
+            currentTimestamp = Instant.now();
+            queue.put(KV.of(message, currentTimestamp));
+          }
+
+          @Override
+          public void deliveryComplete(IMqttDeliveryToken token) {
+            // nothing to do
+          }
+        });
+        return advance();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      LOGGER.info("Taking from the pending queue ({})", queue.size());
+      try {
+        KV<MqttMessage, Instant> message = queue.take();
```

Comment:
Yes, according to the javadoc, `take()` will wait if the blocking queue is empty.